generator/generator.go

```diff
@@ -0,0 +1,163 @@
+package generator
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/helm/helm/log"
+)
+
+const GeneratorKeyword = "helm:generate "
+
+// Walk walks a chart directory and executes generators as it finds them.
+//
+// Returns the number of generators executed.
+func Walk(dir string) (int, error) {
+	count := 0
+	err := filepath.Walk(dir, func(path string, fi os.FileInfo, err error) error {
+
+		// dive-bomb if we hit an error.
+		if err != nil {
+			return err
+		}
+
+		// Skip directory entries. If the directory prefix is . or _, skip the
+		// contents of the directory as well.
+		if fi.IsDir() {
+			return skip(path)
+		}
+
+		f, err := os.Open(path)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		line, err := readGenerator(f)
+		if err != nil {
+			return err
+		}
+		if line == "" {
+			return nil
+		}
+		// Run the generator.
+		os.Setenv("HELM_GENERATE_COMMAND", line)
+		os.Setenv("HELM_GENERATE_FILE", path)
+		os.Setenv("HELM_GENERATE_DIR", dir)
+		line = os.ExpandEnv(line)
+		os.Setenv("HELM_GENERATE_COMMAND_EXPANDED", line)
+		log.Debug("File: %s, Command: %s", path, line)
+		count++
+		err = execute(line)
+		if err != nil {
+			return fmt.Errorf("failed to execute %s (%s): %s", line, path, err)
+		}
+		return nil
+	})
+
+	return count, err
+}
+
+func execute(command string) error {
+	args := strings.Fields(command)
+	if len(args) == 0 {
+		return errors.New("empty command")
+	}
+	name := args[0]
+	args = args[1:]
+
+	cmd := exec.Command(name, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
```

Comment:
I'm actually not sure. The environment variables are used by:

- the generate command itself, which runs substitutions using ExpandEnv
- the generator plugin, which uses environment variables to find out about the environment
- and possibly children of the plugin, which may use those variables for, say, substitution

Since `os.ExpandEnv`, which runs in-process, doesn't seem to be setable to the environment of the `cmd.Env`, I don't think switching would work.